### ISSUE TITLE
Line-length doesn't work

### ISSUE THREAD

**Issue by t-rad679** [markdownlint/markdownlint#295]
## Description
MD013 -- line-length does not seem to flag lines that are too long
### Environment

* Mac OSX 10.15.2
* mdl installed with gem

**MDL Version**
0.8.0

### Expected Behavior
MD013 should activate and report that a line is too long

### Actual Behavior
Got the output below. Notice that there is no report for MD013:

```
test.md:11: MD004 Unordered list style
test.md:12: MD004 Unordered list style
test.md:9: MD005 Inconsistent indentation for list items at the same level
test.md:12: MD007 Unordered list indentation
test.md:6: MD009 Trailing spaces
test.md:13: MD009 Trailing spaces
test.md:22: MD009 Trailing spaces
test.md:23: MD009 Trailing spaces
test.md:2: MD022 Headers should be surrounded by blank lines
test.md:25: MD022 Headers should be surrounded by blank lines
test.md:20: MD029 Ordered list item prefix
test.md:21: MD029 Ordered list item prefix
test.md:8: MD030 Spaces after list markers
test.md:19: MD030 Spaces after list markers
test.md:20: MD030 Spaces after list markers
test.md:21: MD030 Spaces after list markers
```
## Replication Case
Run the following command with the files below:

```bash
mdl -c .markdownlintrc.js test.md
```

#### .markdownlintrc.js

```js
module.exports = {
    "default": true,
    // Rule customizations for markdownlint go here
    // https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md

    "line_length": {
        "line_length": 120,
        "strict": true
    },

    // ===
    // Prettier overrides
    // ===
    "no-multiple-blanks": false,
    "list-marker-space": false,
    "no-inline-html": false
}
```

#### test.md

```md
Lines Before Title
# Title

No TOC

# Multiple Top Level Headers 

*   This uses the wrong unordered list marker
  * This is the wrong indent

- This has the wrong number of spaces after marker
    - asdf
    
This line is tooooooooooooooooooooooooooooooooooooooooooooo looooooooooooooooooooooooooooooooooooooooooooooooooooooooooong.
[link](https://makesurelinkisnotsplitbymaxlinlengthcheckbecauseofthestrictmodeoptioninthelinelengthruleinmarkdownlintrc.com)
Wrong Header style
--------------------------

1.  Non-lazy list
    2.  boop
    3.  beep
   
trailing whitespace       
no blank line before heading
## Heading
No blank line after heading
```

Note that I've tried a number of variations. For example:
* Swap ```line_length``` for ```line-length```
* Setting ```"line-length": true```
* Adding ```"rules": "MD013"```
* Setting ```"default": false"

Lastly, note that mdl does not throw an error when there are invalid rules configured in .markdownlintrc

**Comment by jaymzh** [comment]
I believe you are using the NodeJS version of Markdownlint which is not this software, this is the ruby version. @DavidAnson  ?

**Comment by DavidAnson** [comment]
I wondered about that myself, but the “Environment” section at the top is explicit about being the Ruby version. Can you help us understand, @t-rad679 ?

**Comment by t-rad679** [comment]
Hey guys,

I'm really sorry I dropped off the map. I got pulled onto other work and this got lost in a flood of emails. I do also use the Node version on another project, but it is installed locally to that project and run with yarn. I am quite sure I'm using the Ruby installed version of mdl. in this case.

I do find it strange that the Ruby version is so far behind the Node version. Is that just that the Ruby version was made later and has its own version track?

What else do you need to know?

**Comment by jaymzh** [comment]
The version numbers of the two are entirely unrelated. :) I believe this one was first and the the Node one came along. They're done by entirely different group of people who release on entirely different schedules, with some general effort  to keep the rule numbers in sync.

Anyway - you're passing `.markdownlintrc.js` which is a Javascript file - to `mdl` - which means you **are** using the Javascript version of MDL, because the ruby one cannot parse javascript. Otherwise you'd get parse errors like this:

```
b/config.rb:63:in `instance_eval': /tmp/markdownlintrc.js:1: syntax error, unexpected '.' (SyntaxError)
module.exports = {
      ^
/tmp/markdownlintrc.js:3: syntax error, unexpected tCONSTANT, expecting =>
    // Rule customizations for markdownl...
       ^~~~
/tmp/markdownlintrc.js:3: syntax error, unexpected tIDENTIFIER, expecting do or '{' or '('
...tomizations for markdownlint go here
...                             ^~
/tmp/markdownlintrc.js:4: syntax error, unexpected tIDENTIFIER, expecting end-of-input
    // https://github.com/DavidAnson/mark...
       ^~~~~
```

So - again, if the output you provided is valid, then you are using the Javascript version of MDL in that output. If you're using the ruby version when you ran into this bug, then you provided the wrong output.

**Comment by t-rad679** [comment]
Yeah I've been running into that when I've been running it recently. Maybe in the past I converted it to JSON and for some reason didn't copy the correct command? I copied ```.markdownlintrc.js``` to ```.markdownlint``` and converted it to JSON, then ran the following command just now:

```
mdl test.md
```

where test.md is directly copied from the initial description of the issue. The command I pasted above is also copied directly from the command line where I ran it. The output of the command was:

```
test.md:11: MD004 Unordered list style
test.md:12: MD004 Unordered list style
test.md:9: MD005 Inconsistent indentation for list items at the same level
test.md:12: MD007 Unordered list indentation
test.md:6: MD009 Trailing spaces
test.md:13: MD009 Trailing spaces
test.md:22: MD009 Trailing spaces
test.md:23: MD009 Trailing spaces
test.md:2: MD022 Headers should be surrounded by blank lines
test.md:25: MD022 Headers should be surrounded by blank lines
test.md:20: MD029 Ordered list item prefix
test.md:21: MD029 Ordered list item prefix
test.md:8: MD030 Spaces after list markers
test.md:19: MD030 Spaces after list markers
test.md:20: MD030 Spaces after list markers
test.md:21: MD030 Spaces after list markers
```

Still no output for lines too long. Just to confirm that I used the ruby version just now, I ran the following command in the same terminal window:

```
which mdl
```

with the output:

```
/Users/trevorradcliffe/.rbenv/shims/mdl
```

copied directly from the console.

**Comment by t-rad679** [comment]
I'm running into something very strange now. I'm running ```mdl```  (still the ruby version, for sure) on another file in another directory and, not only is it triggering the line length rule, it's triggering as though the max line length is the default 80 when it is explicitly set to 120 in the config.

Note that I have changed the config file to yaml so I could bring back my comments. I have confirmed that the rule still does not trigger with the original test file even with the yaml config. Here is the yaml version:

```
"default": false

# Rule customizations for markdownlint go here
# https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md
"line_length":
    "line_length": 120
    "strict": true

# ===
# Prettier overrides
# ===
"no-multiple-blanks": false
"list-marker-space": false
"no-inline-html": false
```

I have played around with underscores vs dashes in the line length rule, both in the outer rule definition and the parameter.

**Comment by jaymzh** [comment]
None of that is a valid config...

First the `-c`, i.e. config, is only things you can pass on the command line. You want a `style`, i.e. `-s`. And in that case, you'd follow the [creating a style](https://github.com/markdownlint/markdownlint/blob/master/docs/creating_styles.md) doc, and make something like this:

```
rule 'MD013', :line_length => 80
```

And then, it'll do what you want.

**Comment by t-rad679** [comment]
Hmmm that's odd. I've seen lots of examples on GitHub that are similar. Given what you said, I'm a little confused about the use case for the config file vs the style. Also, we've been using markdownlintrc for a while now and, I guess I could be wrong, but I was under the impression it was working fine previously. Or at least the owners of the code I've started working on never complained.

I'll try out the style and let you know what happens

**Comment by t-rad679** [comment]
Ok, I've created a style file and am having similar results. I did add some additional rules, but the line length should still trigger. Here's the style file:

```ruby
rule 'MD001' # header-incement
rule 'MD002' # first-header-h1
rule 'MD003', :style => 'atx' # header-style: Use # to denote header

# TODO: Talk about using dash vs asterisk
rule 'MD004', :style => 'consistent' # ul-style

rule 'MD005' # list-indent: Enforces consistent list indentation
rule 'MD006' # ul-start-left
rule 'MD007', :indent => 4 # ul-indent
rule 'MD009' # no-trailing-spaces
rule 'MD010' # no-hard-tabs
rule 'MD011' # no-reversed-links
rule 'MD012' # no-multiple-blanks
rule 'MD013', :line_length => 120 # line-length
rule 'MD014' # commands-show-output: Triggers when bash code is preceded by $, unless command output is displayed
rule 'MD018' # no-missing-space-atx
rule 'MD019' # no-multiple-space-atx
rule 'MD022' # blanks-around-headers
rule 'MD023' # header-start-left
rule 'MD024' # no-duplicate-header
rule 'MD025' # single-h1
rule 'MD026' # no-trailing-punctuation: Triggers on punctuation at end of header
rule 'MD027' # no-multiple-space-blockquote
rule 'MD028' # no-blanks-blockquote
rule 'MD029' # ol-prefix: Enforces lazy ordered lists
rule 'MD030', :ul_multi => 3, :ol_multi => 2 # list-marker-space
rule 'MD031' # blanks-around-fences
rule 'MD032' # blanks-around-lists
rule 'MD033' # no-inline-html
rule 'MD034' # no-bare-urls: URLs must be linked
rule 'MD035' # hr-style: Enforces consistent horizontal rule style
rule 'MD036' # no-emphasis-as-header: Triggers on single lines of entirely emphasized text
rule 'MD037' # no-space-in-emphasis
rule 'MD038' # no-space-in-code
rule 'MD039' # no-space-in-link
rule 'MD041' # first-line-h1
rule 'MD046' # code-block-style: Enforces fencing for code blocks (instead of indentation)
```

The command I'm running is: 

```bash
mdl -s style.rb test.md
```

I also tried it with a ```.markdownlint.yaml``` containing only ```default: false``` as the ```-c``` option.

The line length rule is still not triggering.

**Comment by t-rad679** [comment]
I think this may be the culprit:

> This rule has an exception where there is no whitespace beyond the configured line length. This
> allows you to still include items such as long URLs without being forced to break them in the middle.

Why not check for when there are no spaces on the *entire line*, instead of just after the break, or at least provide an option for that? That's the case I want to except. If a contiguous piece of text is too long, I want it to be on its own line.

**Comment by jaymzh** [comment]
So, in general, that seems sane, but it's not *quite* that simple. What if you have a URL in a list-item? There's a space on either side of the list-item-marker. What if it's in a indented code-block, that one's at least easier, because you can trim leading space. So... yeah, I agree, it's not perfect, but I'm not exactly certain the algorithm I'd use to make it better.

One thing we could do is have a few configurable options like:
```
:allow_nonspaced_lines => true, :allow_nonspaced_overflow => true
```

**Comment by DavidAnson** [comment]
The Node implementation has added a few parameters over time: https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md#md013

**Comment by t-rad679** [comment]
Yeah, I see what you mean. I saw the change to the node version for this, but my understanding is that that option is an all or nothing--As in, if you turn on their ```strict``` parameter, it won't allow *any* overflow. I could be wrong though. Also, the node version seems to only work in node projects. I looked into it because they have a ```--fix``` option, but the command line application in another repo doesn't have it.

**Comment by DavidAnson** [comment]
Yes, the meaning of the strict option is all or nothing. As you note, the Node CLI does support fixing some violations. It can be used in any environment where the Node runtime is available, such as most CI.

Here’s a link: https://github.com/igorshubovych/markdownlint-cli

**Comment by t-rad679** [comment]
Hmmm...When I try to use it outside of a project with a ```package.json``` it gives me an error

**Comment by DavidAnson** [comment]
That’s not enough information for me to help, sorry. In general, I would expect that you install a recent version of Node (10 or later), then install the CLI package globally, and then you should be able to run it with the “markdownlint” command. If that’s not working, please share the relevant console output in a new issue in that project and I can try to help further.

**Comment by t-rad679** [comment]
Yeah, it's just that the CLI package doesn't have the ```--fix``` option. The Node version without the CLI package can't be used outside the context of a Node package.

**Comment by t-rad679** [comment]
Oh, I see now that the markdownlint-cli page says it does have ```--fix```. My version doesn't seem to have it. Let me investigate.

**Comment by t-rad679** [comment]
I guess when I installed with ```yarn``` I ended up working with a different version. I did ```npm install --global markdownlint-cli``` and it showed the ```--fix``` option

**Comment by t-rad679** [comment]
Thanks for the help guys! I guess I have my answer. Is there any plan to work on changes to the maximum line length rule to support some of this behavior?

**Comment by jaymzh** [comment]
I'm happy to put the suggestions I made in https://github.com/markdownlint/markdownlint/issues/295#issuecomment-585305643 on the list. Want to file a separate RFE ticket for them and then we can close this out?

**Comment by t-rad679** [comment]
What is RFE? Would that be the enhancement label? I'll go ahead create the ticket and add whatever labels you suggest.

**Comment by jaymzh** [comment]
RFE = "request for enhancement" - just use the `enhancement` label. Like you did. So I'll close this. :)

**Comment by t-rad679** [comment]
Thank you!
